Check for FIXME etc. during CI

rptb1 opened the pull request:
Add a CI check for FIXME so that code with FIXMEs in it can't pass CI checks, proc.review, or proc.merge.  The purpose is to allow devs to use FIXME to mark things that must be fixed in development branches before merging.

See also https://github.com/Ravenbrook/mps/issues/140#issuecomment-1434317186

rptb1 commented:
Executing [proc.review.entry](https://github.com/Ravenbrook/mps/blob/4634a33f18dc6b30675235405ca43a3636f4942b/procedure/review.rst#51-review-entry)

1. Start time 06:42.
2. Change is small and low risk but nobody is available for express review. Plan to batch with other similar changes on 2023-02-23.
3. proc.review.entry.criteria: Applying [entry.universal](https://github.com/Ravenbrook/mps/blob/branch/2023-01-19/review-procedure/procedure/entry.universal.rst) and [entry.impl](https://github.com/Ravenbrook/mps/blob/branch/2023-01-19/review-procedure/procedure/entry.impl.rst)
4. entry.universal.reason: Pull request describes reason, and links to an issue about why there isn't a better documented reason!
5. Entry passed.
6. Entry took 5 mins.

rptb1 commented:
Executing [proc.review.plan](https://github.com/Ravenbrook/mps/blob/4634a33f18dc6b30675235405ca43a3636f4942b/procedure/review.rst#52-review-planning)

1. Start time 06:47.
2. proc.review.plan.time: Change combines two things: a new script with CI, and discharging about 10 FIXMEs.  Each one of the FIXMEs might take a couple of minutes to check.  Total checking time about 30 mins.
3. proc.review.plan.role: @thejayps could check CI and scripts for correctness. @UNAA008 check FIXME changes.  @rptb1 checks convention and clarity.
4. proc.review.plan.schedule: @UNAA008 @thejayps @rptb1 will review on 2023-02-23 at 11:00.
5. proc.review.plan.homework: @thejayps should read https://docs.github.com/en/actions/using-workflows/about-workflows
6. Planning took 10 mins.

rptb1 commented on tool/check-fixme:
```diff
@@ -0,0 +1,82 @@
+#
+# TODO: A much older convention in the MPS (predates the adoption of
+# "FIXME") is to mark such things with "@@@@".  We should find and fix
+# those too, eventually.
```
m. Insert reference to https://github.com/Ravenbrook/mps/issues/162

rptb1 replied:
Fix: Added in 3f46141e3f1ca46f64cb8f9bb3b8e38f715bee8a

rptb1 commented on tool/check-fixme:
```diff
@@ -0,0 +1,82 @@
+
+# Exclude some things that:
+# - we have no control over
+# - come from elsewhere
```
m. rule.generic.clear: elsewhere -> outside the MPS project

rptb1 replied:
Fix: Changed in 3f46141e3f1ca46f64cb8f9bb3b8e38f715bee8a

rptb1 commented on tool/check-fixme:
```diff
@@ -0,0 +1,82 @@
+# - we have no control over
+# - come from elsewhere
+# - build results that are copies of other things
+# - things that legit talk about FIXMEs (like this script)
```
m. rule.generic.self: the command excludes backups too

rptb1 replied:
Fix: Added comment about local editor backups in 3f46141e3f1ca46f64cb8f9bb3b8e38f715bee8a

UNAA008 commented on code/config.h:
```diff
@@ -393,7 +393,7 @@
 
 /* Pool MVT Configuration -- see <code/poolmv2.c> */
-/* FIXME: These numbers were lifted from mv2test and need thought. */
+/* TODO: These numbers were lifted from mv2test and need thought. */
```
This comment may need more explanation about the impact of the number choices.

rptb1 replied:
Raise: #176

UNAA008 commented on code/splay.c:
```diff
@@ -1237,7 +1237,7 @@ Bool SplayFindFirst(Tree *nodeReturn, SplayTree splay,
     Tree oldRoot, newRoot;
 
-    /* FIXME: Rename to "seen" and "not yet seen" or something. */
+    /* TODO: Rename to "seen" and "not yet seen" or something. */
```
For what reason exactly?

rptb1 replied:
There's a general rule (possibly not documented) that we avoid the terms "old" and "new" because they often meaningless and relative. So this might been just that -- find better names specific to the context.  So:

m: TODO should say this.
mI: Add the rule about "old" and "new" probably to rule.impl.

rptb1 replied:
1. Fix: Clarified in 3c2c0797d0e9f8497b72ed0cf79ac2bf3c286c4e
2. Forget: Wasn't that relevant in this case. It'll come up again.

rptb1 commented on code/config.h:
```diff
@@ -393,7 +393,7 @@
 
 /* Pool MVT Configuration -- see <code/poolmv2.c> */
-/* FIXME: These numbers were lifted from mv2test and need thought. */
+/* TODO: These numbers were lifted from mv2test and need thought. */
```
m. This and other constants in config.h are not justified here or in the design.  See https://github.com/Ravenbrook/mps/blob/1d940552cf33ceb46c807288d953e7e308988824/design/config.txt#to-document .  Raise an issue?

rptb1 replied:
@UNAA008 adds: TODOs especially need to be comprehensible to people who come across them without deep knowledge of the thing they annotate.  Form part of the doc describing how to use TODOs.

rptb1 replied:
1. Raise: #176 and also added references to that in 19d7fafaa4924512370bfb95e50a1570caed03b8
2. Raise: https://github.com/Ravenbrook/mps/issues/140#issuecomment-1443203684

rptb1 reviewed (commented):
Executing [proc.review.check](https://github.com/Ravenbrook/mps/blob/4634a33f18dc6b30675235405ca43a3636f4942b/procedure/review.rst#54-review-checking)

1. Start time 11:16
2. 4 minor defects

UNAA008 reviewed (commented):
The general principle followed to convert FIXME to TODO worked well.
I had 2 comments requesting clarificxation inserted inline.

thejayps reviewed (changes requested):
m https://memory-pool-system.readthedocs.io/en/latest/design/guide.developer.html Doesn't talk about the difference between FIXME and TODO

M https://memory-pool-system.readthedocs.io/en/latest/design/guide.review.html seems to be deprecated. Danger that develepers will pass reviews based on this doc


q Where do we state that CI will fail if fixme's are found, will they fail if fixmes are found

q Why is the job in fixme-check.yml called check-rst? what has this to do specifically with restructured text?

rptb1 commented:
Executing [proc.review.log](https://github.com/Ravenbrook/mps/blob/4634a33f18dc6b30675235405ca43a3636f4942b/procedure/review.rst#55-review-logging)

1. Start time 11:31
2. @UNAA008 found 2 minor, @thejayps find 1 M, 1 m,  2 q, @rptb1 4 m.
3.

thejayps commented on tool/check-fixme:
```diff
@@ -0,0 +1,82 @@
+# "FIXME") is to mark such things with "@@@@".  We should find and fix
+# those too, eventually.
+
+# Exclude some things that:
```
m. What's the rule for putting things on this exclusion list?

rptb1 replied:
Fix: Clarified what the list is for in 3f46141e3f1ca46f64cb8f9bb3b8e38f715bee8a

rptb1 commented:
> m https://memory-pool-system.readthedocs.io/en/latest/design/guide.developer.html Doesn't talk about the difference between FIXME and TODO

Raise: Already part of https://github.com/Ravenbrook/mps/issues/140#issuecomment-1434317186 

> M https://memory-pool-system.readthedocs.io/en/latest/design/guide.review.html seems to be deprecated. Danger that develepers will pass reviews based on this doc

Raise: https://github.com/Ravenbrook/mps/pull/123#issuecomment-1443116900

> q Where do we state that CI will fail if fixme's are found, will they fail if fixmes are found

Fix: Updated link in 38d9feb69481f647909263f19b10e704f17fef56

> q Why is the job in fixme-check.yml called check-rst? what has this to do specifically with restructured text?

Fix: Corrected mistake in 38d9feb69481f647909263f19b10e704f17fef56

rptb1 commented:
Executing [proc.review.exit](https://github.com/Ravenbrook/mps/blob/39fbd18bb93145edc2ed2134c861892a54452479/procedure/review.rst#58-review-exit)

1. Start time 12:54.
2. Revised change passed.
3. review.exit.calc:
  - hours used: 90 minutes
  - hours saved: 2 hours fixing review procedure (and the manual)
  - major defects remaining: 0

thejayps reviewed (approved):
proc.review.exit.pass
